README.md
# GrasPlan website

GrasPlan is een digitale graslandgebruikskalender. Met GrasPlan is het mogelijk om voor al je percelen het graslandmanagement bij te houdem. Je kunt hierin aangeven wanneer je gemaaid, beweid of bemest heeft. Daarna heb je direct met GrasPlan een eenvoudig van al percelen. GrasPlan ondersteunt ook rotatieweiden.

GrasPlan wordt in 2020 ook ingezet binnen de GLB-pilot in Noord-Holland. GrasPlan is een eerste prototype. Heb je waardevolle feedback over de applicatie of wensen qua gebruikersvriendelijkheid en functionaliteit, laat het ons weten. Mail je reactie naar info@grasplan.nl of stuur een issue in op GitHub.

Je kunt gebruik maken van GrasPlan op [grasplan.nl](https://grasplan.nl)

## Installatie

De website van GrasPlan is open-source beschikbaar. Om GrasPlan lokaal te draaien moet je eerst NodeJS 12 ([download](https://nodejs.org/en/download/)) hebben geïnstalleerd. Je kunt daarna met de volgende code GrasPlan draaien.

```
git clone https://github.com/AgroCares/GrasPlan-website.git
npm install
npm start
```

Om connectie met de NMI API te maken en data op te vragen en in te laden, is een `.env` bestand nodig in de `grasplan-website` map met de volgende data:

```
GRAS_SESSION_SECRET=[JOUW_SESSION_SECRET]
API_KEY=[JOUW_API_KEY]
API_URL=[URL_VAN_DE_API]
PORT=5621
NODE_ENV=development
```

Toegang tot de NMI API kan worden aangevraagd door een mail te sturen naar `api@grasplan.nl`.

Heb je ideeën of verbeteringen voor code van GrasPlan? Leuk, stuur ze dan in als issue of pull request!

## Documentatie

Op dit moment is er nog geen uitgebreide documentatie voor grasplan-website. Als je vragen hebt leggen we het je graag uit! Maak hiervoor een issue aan of stuur naar `info@grasplan.nl`.


![Logo of NMI](https://media.licdn.com/dms/image/C560BAQEYGcm4HjNnxA/company-logo_200_200/0?e=2159024400&v=beta&t=u40rJ7bixPWB2SAqaj3KCKzJRoKcqf0wUXCdmsTDQvw)

![eu](https://user-images.githubusercontent.com/37927107/78378810-cd3a8300-75d1-11ea-8948-45f8f845ddc3.jpg)

CHANGELOG.md
# Changelog GrasPlan-website

## Versie 1.5.8 2020-12-04
### Fixed
* De kalender laat nu weer maaien zien wanneer er meerdere maaien waren opgeslagen voor een zone

## Versie 1.5.7 2020-12-01
### Fixed
* Lost een aantal situaties op waarbij de `session_id` niet goed herkend werd

## Versie 1.5.6 2020-11-23
### Fixed
* De kalender doet het weer voor gevallen waarbij een maaien een onbekende datum heeft

## Versie 1.5.5 2020-11-18
### Fixed
* De kalender is verlengd tot 31 november

## Versie 1.5.4 2020-11-05
### Fixed
* Maaidatum werd niet opgeslagen

## Versie 1.5.3 2020-10-26
### Changed
* De kalendar loopt nu door tot 31 oktober ipv 10 oktober

### Fixed
* Kalender liet nu niet alle details zien bij beweiden

## Versie 1.5.2 2020-10-23
### Added
* Laat bij adviseurs de naam van het bedrijf zien boven de kalender

### Fixed
* Fix om `farm_id` weg te halen uit localStorage bij login, anders problemen met meerdere accounts
* Fix wanneer cattle count niet bekend is

## Versie 1.5.1 2020-10-22
### Fixed 
* De functie `setFarmId` werd niet met await aangeroepen, terwijl dat wel nodig is
* Fix om bemesting toe te voegen
* Laat veldnaam zien bij percelen

## Versie 1.5.0 2020-10-21
### Changed
* De backend is nu bijgewerkt naar de laatste versie van de NMI API

## Versie 1.4.1 2020-10-07
### Fixed
* Adviseurs kunnen nu kalenders zien die graslandvernieuwing bevatten
* Adviseurs zien nu een groene streep op de kalender als weiden 1 dag duurt

## Versie 1.4.0 2020-05-04
### Changed
* Bij de kalender worden de zones nu alfabetisch gesorteerd

## Versie 1.3.2 2020-04-22
### Added
* Een korte uitleg over zones is toegevoegd bij de keuze of zones gebruikt of niet
* Bij het toevoegen van beheermaatregelen is een stuk tekst toegevoegd die vertelt dat deze maatregelen bedoeld zijn de GLB-pilot

## Versie 1.3.1 2020-04-14
### Added
* Bij het toevoegen van een actie is er nu een knop om alle percelen/zones te deselecteren

### Changed
* De mobiele versie van de kalendar laat niet meer een template tabel zien, maar de tekst dat de mobiele versie voor de kalendar nog niet beschikbaar is

### Fixed
* In de kalender verschijnt nu wel een streepje wanneer er 1 dag wordt beweid

## Versie 1.3.0 2020-04-10
### Added
* Als een gebruiker in de kalender klikt op een actie, dan verschijnen er details over die actie in een pop-up
* Als een adviseur klikt op een actie in de kalender van een van zijn klanten dan zijn er details te zien over die actie
* Het is mogelijk om een beheersmaatregel of graslandvernieuwing te verwijderen door de gebruiker
* Bij toevoegen van bemesting wordt nu gecontroleerd of een product is geselecteerd en anders verschijnt er een foutmelding daarover
* Bij toevoegen van bemesting wordt nu gecontroleerd of een hoeveelheid is ingevuld en anders verschijnt er een foutmelding daarover
* Bij toevoegen van een beheersmaatregel wordt nu gecontroleerd of een maatregel is geselecteerd en anders verschijnt er een foutmelding daarover

### Changed
* De foutmeldingen bij het toevoegen van een actie zijn wat netter geformuleerd

### Fixed
* Graslandvernieuwing liet in de kalender een verkeerde datum zien.

## Version 1.2.1 2020-04-03
### Added
* Een stukje tekst is toegevoegd dat GrasPlan mede is gefinancieerd door het EU Landbouwfonds

### Changed
* Bij het toevoegen van een actie staan de datums als default op vandaag
* Bij het toevoegen van een bemesting, is het verplicht een mestproduct aan te geven

### Fixed
* Bij het opsturen van een bemesting ging soms wat fout doordat het data object niet goed werd aangemaakt
* Fix security issue GHSA-7fhm-mqm4-2wp7s
* Fix dat de bij eendaagse actie er voor 2 dagen instond #2

## Version 1.2.0 2020-04-03
GrasPlan website is released open-source :tada:

public/js/calendar.js
$(document).ready(async function () {
  M.AutoInit();

  await setFarmId()

  //  Define farm in global env
  let farm_select = await axios({
    method: 'post',
    url: '/api_farm_sel',
    data: {
      farm_id: localStorage.farm_id
    }
  });
  let fields = farm_select.data.data.data.field;

  // ga naar time table als klein scherm
  if (screen.width <= 700) {
    window.location.href = "time_table";
  } else {
    make_timeline(fields);
  }

  timeline_plotly.on('plotly_click', function (e) {

    let actie = e.points["0"].data;

    // voeg info to aan modal
    modalEvent(actie);

  });


  // voeg event toe aan navbar
  $('#navbar').append('<ul class="right"> <li><a id="to_actie" href="create_event" class="waves-effect waves-light btn">Actie toevoegen <i class="material-icons left">add</i></a></li></ul>')

});

modalEvent = function (actie) {

  $("#event_mod_head").empty();
  $('#modalcontent').empty();
  $('#modalcontent').append('<h5> ' + actie.actie + '</h5>');
  $('#modalcontent').append('<p>' + actie.name + '</p>');
  $("#event_modal").modal("open");

  $("#delete_event").on("click", function (e) {
    let eventId = actie.id;
    let eventType = actie.actie;
    deleteEvent(eventType, eventId);
  });

};


deleteEvent = function (eventType, eventId) {

  // if grazing
  if (eventType == 'Beweiden') {

    axios({
      method: "post",
      url: "/api_grazing_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete grazing succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete grazing");
          let text = "Fout bij verwijderen weiden.";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });
  } else if (eventType == 'Maaien') {
    axios({
      method: "post",
      url: "/api_mowing_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete mowing succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete mowing");
          let text = "Fout bij verwijderen maaien.";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });
  } else if (eventType == 'Bemesten') {
    axios({
      method: "post",
      url: "/api_fertilising_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete bemesten succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete fertilizing");
          let text = "Fout bij verwijderen bemesten.";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });
  } else if (eventType == 'Pesticidegebruik') {
    axios({
      method: "post",
      url: "/api_pesticidation_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete pesticide succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete api_pesticidation_delete");
          let text = "Fout bij verwijderen pesticidegebruik.";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });
  } else if (eventType == 'Beheermaatregelen') {
    axios({
      method: "post",
      url: "/api_nature_measure_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete nature measures succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete api_nature_measures_delete");
          let text = "Fout bij verwijderen beheersmaatregel";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });

  } else if (eventType == 'Graslandvernieuwing') {
    axios({
      method: "post",
      url: "/api_grassland_renewal_delete",
      data: { eventId: eventId }
    })
      .then(function (res) {

        if (res.data.success) {
          // Successful response
          console.log("delete grassland renewal succesfull");
          window.location.href = "calendar";

        } else {
          // Unsuccessful response
          console.log("Unsuccesfull delete api_grassland_renewal_delete");
          let text = "Fout bij verwijderen graslandvernieuwing";
          M.toast({ html: text });
        }
      })
      .catch(function (err) {
        console.log(err);
      });

  } else {
    console.log("onbekende actie");
  }

};

function formatDate(date, add) {
  let d = new Date(date);
  d.setDate(d.getDate() + add);
  let month = '' + (d.getMonth() + 1);
  let day = '' + d.getDate();
  let year = d.getFullYear();

  if (month.length < 2)
    month = '0' + month;
  if (day.length < 2)
    day = '0' + day;

  return [year, month, day].join('-');
}

function timestamp(str) {
  return new Date(str).getTime();
}



make_timeline = function (fields) {
  var container = document.getElementById("timeline_plotly");

  var color_maaien = "rgb(204, 0, 0)";
  var color_weiden = "rgb(0, 204, 0)";
  var color_bemesten = "rgb(102, 26, 0)";
  var color_pesticide = "rgb(255, 204, 0)";
  var color_invisible = "rgba(0,0,0,0)";
  var color_nature = "rgb(85, 205, 255)";
  var color_renewal = "#034d02";
  var color_background = "#efebe9";

  // define zone list
  let zone_list = [];
  fields.forEach(field => {
    // console.log(field.zone[0].zon_id)
    if (field.zone.length == 1) {
      let zone_info = {
        name: field.field_name,
        id: field.zone[0].zone_id
      };
      zone_list.push(zone_info);
    } else {
      field.zone.forEach(zone => {
        if (zone.zone_name != null) {
          let zone_info = {
            name: field.field_name + " (" + zone.zone_name + ")",
            id: zone.zone_id
          }
          zone_list.push(zone_info);
        }
      });
    }
  });

  //Sort zone_list alphabetically
  zone_list.sort((a, b) => (a.color > b.color) ? 1 : -1);

  // Define data list with actions
  let actions = [];

  // set up legend
  let data_plot = [
    {
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_maaien, "width": 20 },
      mode: 'lines',
      name: 'Maaien',
      showlegend: true
    },
    {
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_weiden, "width": 20 },
      mode: 'lines',
      name: 'Weiden',
      showlegend: true
    },
    {
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_bemesten, "width": 20 },
      mode: 'lines',
      name: 'Bemesten',
      showlegend: true
    },
    {
      x: ["2020-01-10", "2020-01-12"],
      perceel: 0,
      line: { "color": color_pesticide, "width": 20 },
      mode: 'lines',
      name: 'Pesticidegebruik',
      showlegend: true
    },
    {
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_nature, "width": 20 },
      mode: 'lines',
      name: 'Beheermaatregel',
      showlegend: true
    },
    {
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_renewal, "width": 20 },
      mode: 'lines',
      name: 'Graslandvernieuwing',
      showlegend: true
    },
    { // overwrite colors with blank :D 
      x: ["2020-01-10", "2020-01-10"],
      perceel: 0,
      line: { "color": color_background, "width": 25 },
      mode: 'lines',
      showlegend: false
    },

  ];

  fields.forEach(field => {
    field.zone.forEach(zone => {
      // events for mowing
      if (zone.mowing.length > 0) {
        zone.mowing.forEach(item => {
          if (item.mowing_date != null) {
            actions.push({
              x0: formatDate(item.mowing_date, 0),
              x1: formatDate(item.mowing_date, 1),
              y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
              y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
              line: { width: 0 },
              // type: "rect",
              fillcolor: color_maaien
            });
            data_plot.push({
              actie: 'Maaien',
              id: item.mowing_id,
              x: [formatDate(item.mowing_date, 0), formatDate(item.mowing_date, 0)],
              y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
              perceel: zone_list.findIndex(x => x.id == zone.zone_id),
              marker: { "color": color_invisible },
              name: '<b>Maaidatum:</b> ' + formatDate(item.mowing_date, 0),
              text: 'Maaien',
              hoverinfo: "x+text",
              uid: "c2e171",
              showlegend: false
            });
          }
        });
      }
      // events for grazing
      if (zone.grazing.length > 0) {
        zone.grazing.forEach(item => {
          if (item.cattle_type == null) {
            item.cattle_type = "Onbekend";
          }
          if (item.cattle_count == null) {
            item.cattle_count = "Onbekend";
          }
          let x1 = formatDate(item.grazing_end_date, 0)
          if (item.grazing_end_date == item.grazing_start_date) {
            x1 = formatDate(item.grazing_end_date, 1);
          }
          actions.push({
            // actie: 'weiden',
            x0: formatDate(item.grazing_start_date, 0),
            x1: x1,
            y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
            y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
            line: { width: 0 },
            fillcolor: color_weiden
          });
          data_plot.push({
            actie: 'Beweiden', // even voor de console
            id: item.grazing_id,
            x: [formatDate(item.grazing_start_date, 0), formatDate(item.grazing_end_date, 0)],
            y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
            perceel: zone_list.findIndex(x => x.id == zone.zone_id),
            marker: { "color": color_invisible },
            name: '<b>Begindatum beweiden:</b> ' + formatDate(item.grazing_start_date, 0) + '<br><b>Einddatum beweiden:</b> ' + formatDate(item.grazing_end_date, 0) + '<br><b>Diergroep:</b> ' + item.cattle_type + '<br><b>Aantal dieren:</b> ' + item.cattle_count,
            text: 'Beweiden',
            hoverinfo: "x+text",
            uid: "c2e171",
            showlegend: false
          });
        });

      }
      // events for fertilizing 
      if (zone.fertilization.length > 0) {
        zone.fertilization.forEach(item => {
          if (item.fertilizer_name == null) {
            item.fertilizer_name = "Onbekend";
          }
          if (item.fertilization_amount == null) {
            item.fertilization_amount = "Onbekend";
          }
          actions.push({
            x0: formatDate(item.fertilization_date, 0),
            x1: formatDate(item.fertilization_date, 1),
            y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
            y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
            line: { width: 0 },
            // type: "rect",
            fillcolor: color_bemesten
          });
          data_plot.push({
            actie: 'Bemesten',
            id: item.fertilization_id,
            x: [formatDate(item.fertilization_date, 0), formatDate(item.fertilization_date, 0)],
            y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
            perceel: zone_list.findIndex(x => x.id == zone.zone_id),
            marker: { "color": color_invisible },
            name: '<b>Bemestingsdatum:</b> ' + formatDate(item.fertilization_date, 0) + '<br><b>Bemestingsproduct:</b> ' + item.fertilizer_name + '<br><b>Bemestingshoeveelheid:</b> ' + item.fertilization_amount + ' kg per hectare',
            text: 'Bemesten',
            hoverinfo: "x+text",
            uid: "c2e171",
            showlegend: false
          });
        });
      }
      // events for pesticide 
      if (zone.pesticidation.length > 0) {
        zone.pesticidation.forEach(item => {
          actions.push({
            x0: formatDate(item.pesticidation_date, 0),
            x1: formatDate(item.pesticidation_date, 1),
            y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
            y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
            line: { width: 0 },
            // type: "rect",
            fillcolor: color_pesticide
          });
          data_plot.push({
            actie: 'Pesticidegebruik',
            id: item.pes_id,
            x: [formatDate(item.pesticidation_date, 0), formatDate(item.pesticidation_date, 0)],
            y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
            perceel: zone_list.findIndex(x => x.id == zone.zone_id),
            marker: { "color": color_invisible },
            name: '<b>Datum van pesticidegebruik:</b> ' + formatDate(item.pesticidation_date, 0),
            text: 'Pesticidegebruik',
            hoverinfo: "x+text",
            uid: "c2e171",
            showlegend: false
          });
        });
      }
      // Events for nature management
      if (field.nature_management.length > 0) {
        zone.nature_management = [];
        field.nature_management.forEach(item => {
          actions.push({
            x0: formatDate(item.nature_management_date, 0),
            x1: formatDate(item.nature_management_date, 1),
            y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
            y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
            line: { width: 0 },
            // type: "rect",
            fillcolor: color_nature
          });
          data_plot.push({
            actie: 'Beheermaatregelen',
            id: item.nature_management_id,
            x: [formatDate(item.nature_management_date, 0), formatDate(item.nature_management_date, 0)],
            y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
            perceel: zone_list.findIndex(x => x.id == zone.zone_id),
            marker: { "color": color_invisible },
            name: '<b>Datum van beheermaatregel:</b> ' + formatDate(item.nature_management_date, 0) + '<br><b>Beheermaatregel:</b> ' + item.nature_measure_name,
            text: 'Beheermaatregelen',
            hoverinfo: "x+text",
            uid: "c2e171",
            showlegend: false
          });
        });
      }
      // Events for nature management
      if (field.grassland_renewal.length > 0) {
        zone.grassland_renewal = [];
        field.grassland_renewal.forEach(item => {
          actions.push({
            x0: formatDate(item.grassland_renewal_date, 0),
            x1: formatDate(item.grassland_renewal_date, 1),
            y0: zone_list.findIndex(x => x.id == zone.zone_id) - 0.4,
            y1: zone_list.findIndex(x => x.id == zone.zone_id) + 0.4,
            line: { width: 0 },
            // type: "rect",
            fillcolor: color_renewal
          });
          data_plot.push({
            actie: 'Graslandvernieuwing',
            id: item.nma_id,
            x: [formatDate(item.grassland_renewal_date, 0), formatDate(item.grassland_renewal_date, 0)],
            y: [zone_list.findIndex(x => x.id == zone.zone_id) - 0.4, zone_list.findIndex(x => x.id == zone.zone_id) + 0.4],
            perceel: zone_list.findIndex(x => x.id == zone.zone_id),
            marker: { "color": color_invisible },
            name: '<b>Datum van graslandvernieuwing:</b> ' + formatDate(item.grassland_renewal_date, 0),
            text: 'Graslandvernieuwing',
            hoverinfo: "x+text",
            uid: "c2e171",
            showlegend: false
          });
        });
      }
    });
  });

  // zone names to plot 
  let names = zone_list.map(item => item.name);

  // tijdelijk: voor elk veld sowieso 1 actie toevoegen met niets er in, anders komen niet alle velden in calender
  fields.forEach(field => {
    field.zone.forEach(zone => {
      actions.push({
        x0: '20-03-01',
        x1: '20-03-01',
        y0: zone_list.findIndex(x => x.id == zone.zone_id) + 1 - 0.4,
        y1: zone_list.findIndex(x => x.id == zone.zone_id) + 1 + 0.4,
        line: { width: 0 },
        type: "rect",
        opacity: 1,
        fillcolor: color_maaien
      });
      data_plot.push({
        actie: 'maaien',
        x: ['20-03-01', '20-03-01'],
        perceel: zone_list.findIndex(x => x.id == zone.zone_id) + 1,
        "marker": { "color": color_invisible },
        showlegend: false
      });
    });
  });

  // Layout of plotly
  var layout = {
    xaxis: {
      type: "date",
      side: "top",
      showgrid: true,
      gridcolor: "#bdbdbd",
      zeroline: false,
      tick0: '2020-01-01',
      // dtick: 86400000.0 * 7,
      tickformat: '%d/%m',
      // rangeselector: selectorOptions,
      // tickformatstops: ticks,
      // rangeslider: {range: ['2020-01-01', '2021-01-01']},
      range: ['2020-02-20', '2020-11-31'],
      fixedrange: true
    },
    yaxis: {
      showgrid: false,
      ticktext: names,
      tickvals: Array.from(Array(names.length).keys()),
      zeroline: false,
      fixedrange: true
      // autorange: true
    },
    paper_bgcolor: "rgba(0,0,0,0)",
    plot_bgcolor: "rgba(0,0,0,0)",
    margin: { l: 150, r: 60, b: 50, t: 50, pad: -20 },
    height: (names.length + 3) * 35, // dynamic height (1 line = ca 50 pixels)
    shapes: actions,
    hovermode: "closest",
    hoverlabel: { bgcolor: "#FFF" },
    autosize: true,
    legend: { "orientation": "h", y: 0, font: { size: 18 } },
    font: { size: 14 }
  };

  var config = { responsive: true, displayModeBar: false };

  Plotly.newPlot("timeline_plotly", data_plot, layout, config);
};
